Restructure profiling

yinan1048576 opened the pull request:
Develop new data structure and code logic for holding profiling
related information stored in the extent that may be needed after the
extent is released, which in particular is the case for the
reallocation code path (e.g. in `rallocx()` and `xallocx()`).  The
data structure is a generalization of `prof_tctx_t`: we previously
only copy out the `prof_tctx` before the extent is released, but we
may be in need of additional fields. Currently the only additional
field is the allocation time field, but there may be more fields in
the future.

The restructuring also resolved a bug: `prof_realloc()` mistakenly
passed the new `ptr` to `prof_free_sampled_object()`, but passing in
the `old_ptr` would crash because it's already been released.  Now
the essential profiling information is collectively copied out early
and safely passed to `prof_free_sampled_object()` after the extent is
released.

yinan1048576 commented:
This is the real fix following #1685.

yinan1048576 commented on include/jemalloc/internal/arena_inlines_b.h:
```diff
@@ -40,23 +40,29 @@ arena_choose_maybe_huge(tsd_t *tsd, arena_t *arena, size_t size) {
+	assert(prof_info != NULL);
 
 	/* Static check. */
 	if (alloc_ctx == NULL) {
```
It seems to me that this would end up making the logic more complicated, mainly due to:
- We need to define a local `alloc_ctx` variable if `alloc_ctx` is not given.
- We need to avoid the `iealloc()` call when `alloc_ctx` is given and the `slab` field is `true`, and we need to avoid calling `iealloc()` twice when `alloc_ctx` is not given but `extent_slab_get(extent)` turns out to be `false`.

The code would end up being something like:
```
        const extent_t *extent;
        alloc_ctx_t alloc_ctx_local;

        if (alloc_ctx == NULL) {
                extent = iealloc(tsdn, ptr);
                alloc_ctx_local.slab = extent_slab_get(extent);
                alloc_ctx = &alloc_ctx_local;
        } else if (unlikely(!alloc_ctx->slab)) {
                extent = iealloc(tsdn, ptr);
        }

        if (unlikely(!alloc_ctx->slab)) {
                large_prof_info_get(tsdn, extent, prof_info);
        } else {
                prof_info->prof_tctx = (prof_tctx_t *)(uintptr_t)1U;
        }
```
Any idea how to make it simpler?

interwq replied:
It looks we only need the `is_slab` part of the alloc_ctx.

Maybe like this:

```
extent = iealloc(tsdn, ptr);
bool is_slab = alloc_ctx != NULL ? alloc_ctx->slab : extent_slab_get(extent);
```

yinan1048576 replied:
It works other than that `iealloc()` is always called, and the call is unnecessary in the common case i.e. when `alloc_ctx != NULL && alloc_ctx->slab`.

How about:
```
const extent_t *extent = alloc_ctx != NULL && alloc_ctx->slab ? NULL : iealloc(tsdn, ptr);
/* extent is guaranteed to be non-NULL at this point if alloc_ctx != NULL. */
bool is_slab = alloc_ctx != NULL ? alloc_ctx->slab : extent_slab_get(extent);
```
(This code feels pretty tricky though. :) )

yinan1048576 replied:
Sorry in the comment I meant `if alloc_ctx == NULL`.

interwq replied:
Hmm this should work:

```
extent_t *e;
bool is_slab;
if (alloc_ctx != NULL) { 
    slab = alloc_ctx->slab;
    if (!slab) { e = iealloc(tsdn, ptr); }
} else {
    e = iealloc(tsdn, ptr);
    slab = extent_slab_get(extent);
}
if (unlikely(!slab)) {
    // compiler will warn if it find e is used uninitialized.
    large_prof_info_get(tsdn, e, prof_info);
} else {
    prof_info->prof_tctx = (prof_tctx_t *)(uintptr_t)1U;
}

yinan1048576 commented on test/unit/prof_tctx.c:
```diff
@@ -15,8 +15,8 @@ TEST_BEGIN(test_prof_realloc) {
 	assert_ptr_not_null(p, "Unexpected mallocx() failure");
-	tctx_p = prof_tctx_get(tsdn, p, NULL);
-	assert_ptr_ne(tctx_p, (prof_tctx_t *)(uintptr_t)1U,
+	prof_info_get(tsdn, p, NULL, &prof_info_p);
```
Should the setter functions also be changed so as to be consistent?

interwq replied:
If there's no complication (like a chain of change required), I'd updating the setter as well.

yinan1048576 commented on include/jemalloc/internal/prof_inlines_b.h:
```diff
@@ -203,19 +198,20 @@ prof_realloc(tsd_t *tsd, const void *ptr, size_t usize, prof_tctx_t *tctx,
 	 */
 	if (unlikely(old_sampled)) {
-		prof_free_sampled_object(tsd, ptr, old_usize, old_tctx);
+		prof_free_sampled_object(tsd, old_usize, old_prof_info);
```
This is the place fixing the bug.

yinan1048576 commented:
Revised according to feedback. Two extra changes:
- Updated the style of `ixallocx_prof()` in addition to `irallocx_prof()`.
- Added a `memset()` call for `prof_info` when the extent is a slab (in addition to setting the `prof_tctx` field of it), just to avoid any potential UB.

yinan1048576 commented:
Made a separate commit on top to switch `tsdn` to `tsd`. I started with changing the profiling inline getters and setters and ended up with the following changes in addition:
- Switch `tsdn` to `tsd` for `prof_malloc()` and `prof_malloc_sample_object()`, which are necessary.
- Switch `tsdn` to `tsd` for the arena level getters and setters for the profiling fields.
- Removed the `tsdn` input for the large level getters and setters since they're not needed anyway.

yinan1048576 commented:
Verified that `./run_tests.sh` passed.
